archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,264 @@
+(∃ (l : ℤ), a + c = l * l) ∧ (∃ (m : ℤ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  have h := exists_numbers_in_interval n p,
+  cases h with l hl,
+  use (2 * l * l - 4 * l),
+  use (2 * l * l + 1),
+  use (2 * l * l + 4 * l),
+  refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+  { rw le_sub_iff_add_le,
+    exact hl.2.1 },
+  { nlinarith },
+  { nlinarith },
+  { exact hl.2.2},
+  { use (2 * l - 1),
+    ring_exp_eq },
+  { use (2 * l),
+    ring_exp_eq },
+  { use(2 * l + 1),
+    ring_exp_eq },
+    --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+    --construction of a triplet {a, b, c}, which is constructed by setting l=9
+    --in the argument above this lemma.
+  { use 126,
+    use 163,
+    use 198,
+    repeat {split, linarith},
+    refine ⟨_,⟨_,_⟩⟩,
+    use 17,
+    linarith,
+    use 18,
+    linarith,
+    use 19,
+    linarith },
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+(t : A ⊆ finset.range(2 * n + 1)\finset.range n) (h₁ : (n : ℤ) ≤ a)
+(h₂ : (a : ℤ) < b) (h₃ : (b : ℤ) < c) (h₄ : (c : ℤ) ≤ 2 * n) :
+(a ∉ A ↔ a ∈ (finset.range (2 * n + 1)\finset.range n)\A)
+∧ (b ∉ A ↔ b ∈ (finset.range (2 * n + 1)\finset.range n)\A) ∧
+(c ∉ A ↔ c ∈ (finset.range (2 * n + 1)\finset.range n)\A) :=
+begin
+  norm_cast at h₁,
+  norm_cast at h₂,
+  norm_cast at h₃,
+  norm_cast at h₄,
+  refine ⟨⟨_,_⟩, ⟨⟨_,_⟩,⟨_,_⟩⟩⟩,
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,ha⟩,
+    repeat {linarith}},
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range] at ha,
+    exact ha.2 },
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,hb⟩,
+    repeat {linarith}},
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range] at hb,
+    exact hb.2 },
+   {intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range],
+    refine ⟨⟨_,_⟩,hc⟩,
+    repeat {linarith}},
+  { intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range] at hc,
+    exact hc.2 },
+end
+
+
+theorem IMO_2021_Q1 : ∀ (n : ℕ), n ≥ 100 → ∀ (A ⊆ finset.range(2 * n + 1)\finset.range n),
+(∃ (a b : A), a ≠ b ∧ ∃ (k : ℤ), (a : ℤ) + b = k * k) ∨
+(∃ (a b: (finset.range (2 * n + 1)\finset.range n)\A), a ≠ b ∧ ∃(k : ℤ), (a : ℤ) + b=k * k) :=
+begin
+  intros n hn A hA,
+  -- There exists a pairwise unequal triplet a, b, c ∈ [n, 2n]
+  --such that all pairwise sums are perfect squares.
+  have p := exists_triplet_summing_to_squares n hn,
+  rcases p with ⟨a, b, c, ⟨han, hba, hcb, hnc, p⟩⟩,
+  lift a to ℕ,
+  lift b to ℕ,
+  lift c to ℕ,
+  -- Consider by cases based on whether each of {a, b, c} belongs to the 'first pile' A or not.
+  -- In each case, we can find two members of the triplet in the same pile, as required.
+  by_cases h₁: a ∈ A,
+  by_cases h₂: b ∈ A,
+  left,
+  use a,
+  exact h₁,
```

Comment:
```suggestion
  exact or.inl ⟨a, h₁,⟩,
```

Reply:
This fails, at least for me, giving a type class error